### PR TITLE
[SCF] Fixed epilogue predicates in loop pipelining

### PR DESCRIPTION
**Pull request by sjw36**
The computed loop iteration is zero based, so only check it is less than zero. 
This fixes the case when lower bound is not zero.

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp [hunk 1]
```diff
@@ -671,9 +674,9 @@ LoopPipelinerInternal::emitEpilogue(RewriterBase &rewriter,
     setValueMapping(forOp.getInductionVar(), newlastIter, maxStage - i);
 
     if (dynamicLoop) {
-      // pred = iterI >= lb
+      // pred = iterI < 0
       predicates[i + 1] = rewriter.create<arith::CmpIOp>(
-          loc, arith::CmpIPredicate::sge, iterI, lb);
+          loc, arith::CmpIPredicate::slt, iterI, zero);
```
Comment (ThomasRaoux) [inline mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:679]: shouldn't this be `pred = iterI >= 0`? So the predicate is true only when iterI is negative seems odd considering `iterI = total_iters - 1 - i`?

Comment (sjw36) [reply mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:679]: Ah, that's right.
So for:
```
scf.for (lb=3..ub=4, step=2) max_stage = 2                                                        
```
This computes:
```
  total_iterations = (4 - 3) + 2 - 1 / 2 = 1                                                    
  iterI(0) = total_iterations - 1 - 0 = 0                                                       
  newlastIter(0) = iterI(0) * 2 + 3 = 3                                                         
  pred(0) = iterI(0) >= 0 = true                                                                
  iterI(1) = total_iterations - 1 - 1 = -1                                                      
  newlastIter(1) = iterI(1) * 2 + 3 = 1                                                         
  pred(1) = iterI(1) >= 0 = false
```
I'll update now.

**Comment by ThomasRaoux** [comment]
can we add a test or modify one to not start from 0 so that this can be tested?

**Comment by sjw36** [comment]
@ThomasRaoux updated tests.

File: mlir/test/Dialect/SCF/loop-pipelining.mlir [hunk 2]
```diff
@@ -802,8 +801,10 @@ func.func @stage_0_value_escape(%A: memref<?xf32>, %result: memref<?xf32>, %ub:
 //        CHECK:   } else {
 //        CHECK:   }
 //        CHECK:   return
-func.func @dynamic_loop(%A: memref<?xf32>, %result: memref<?xf32>, %lb: index, %ub: index, %step: index) {
+func.func @dynamic_loop(%A: memref<?xf32>, %result: memref<?xf32>, %ub: index) {
   %cf = arith.constant 1.0 : f32
+  %lb = arith.constant 3 : index
+  %step = arith.constant 2 : index
```
Comment (ThomasRaoux) [inline mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: why do we want those to be constants?

Comment (sjw36) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: Just added to make a lower_bound non-zero. The loop is still dynamic with the upper_bound passed in.

Comment (ThomasRaoux) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: wasn't it already a lower_bound non-zero?

Comment (sjw36) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: It helps demonstrates that the `iterI >= 0` is not related to the lower_bound.

Comment (ThomasRaoux) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: how can it be the lower bound if the lower bound is a function argument?

Comment (sjw36) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: %lb and %step were removed from the func args.

Comment (ThomasRaoux) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: my question is how does changing %ld to be a constant helps demonstrate that iterI >= 0 predicate is not related to the lower_bound.

Comment (sjw36) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: I can revert that part (lb, step -> constants) but it should cover the dynamic loop with epilogue peeling either way.

Comment (ThomasRaoux) [reply mlir/test/Dialect/SCF/loop-pipelining.mlir:807]: yeah I would remove it. I think it is good to test the general case where nothing is known about the loop bound

**Comment by sjw36** [comment]
@ThomasRaoux does this look adequate?

**Review by antiagainst — APPROVED** [review]

**Pull request merged**